### PR TITLE
Allow network plugin to expose information about configured network(s) to target container(s)

### PR DESCRIPTION
**Pull request by nickethier**
Its desirable to provide containers with metadata about the networks they are configured for without depending on the kuberenetes-ro api., especially when the network configuration doesn't allow connectivity to the kube apiserver. With this pull the docker manager will set the `NETWORK` environment variable to the last line of the output from the network plugin.

### CODE REVIEW

**Comment by alex-mohr** [comment]
@dchen1107 either yourself or someone on the node team to review, perhaps?

**Comment by dchen1107** [comment]
cc/ @vishh

File: pkg/kubelet/network/exec/exec.go [hunk 1]
```diff
@@ -120,10 +120,15 @@ func (plugin *execNetworkPlugin) validate() error {
...
 
-func (plugin *execNetworkPlugin) SetUpPod(namespace string, name string, id kubeletTypes.DockerID) error {
+func (plugin *execNetworkPlugin) SetUpPod(namespace string, name string, id kubeletTypes.DockerID) (*network.SetUpPodResult, error) {
 	out, err := utilexec.New().Command(plugin.getExecutable(), setUpCmd, namespace, name, string(id)).CombinedOutput()
 	glog.V(5).Infof("SetUpPod 'exec' network plugin output: %s, %v", string(out), err)
-	return err
+	lines := strings.Split(string(out), "\n")
+	result := &network.SetUpPodResult{}
+	if len(lines) > 0 {
+		result.Output = lines[len(lines)-1]
```
Comment (vishh) [inline pkg/kubelet/network/exec/exec.go]: Just curious: Why not return all the lines?

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -64,6 +64,8 @@ const (
 	// we want to able to consider SRV lookup names like _dns._udp.kube-dns.default.svc to be considered relative.
 	// hence, setting ndots to be 5.
 	ndotsDNSOption = "options ndots:5\n"
+
+	containerNetworkEnvKey = "NETWORK"
```
Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: If we ever envision updates to network configurations, an ENV variable is not the best way to expose network information. Has the methodology for introspecting network config been discussed already with the community? 
@thockin @bgrant0607

**Comment by eparis** [comment]
@rajatchopra

**Comment by eparis** [comment]
@dcbw @danwinship

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (rajatchopra) [inline pkg/kubelet/dockertools/manager.go]: How about providing an api at the kubelet itself that the pods can query?
The kubelet can fill up network info about a pod through the status hook in an organized structure. See #9646.

File: pkg/kubelet/network/exec/exec.go (hunk 1 above)
Comment (nickethier) [inline pkg/kubelet/network/exec/exec.go]: Its an irrelevant implementation detail of the network plugin I use. It logs a bit while setting up the network, which is great when debugging but should probably be behind a -debug flag.

I can't think of a good reason it *should* just use the last line so I can fix that if desired.

**Comment by nickethier** [comment]
CLA should be associated with a corporate cla, Jive Communications Inc.

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (eparis) [inline pkg/kubelet/dockertools/manager.go]: Maybe like the kubernetes (and old kubernetes-ro) service we could have something statically defined in the api server. The kube-proxy could then just forward to the kubelet on localhost. Does mean you would need DNS to get the service IP (or use the ENV var, but I know the project as a whole seems to be steering away from ENV vars), to ask about your network configuration. Not sure if that's a problem in anyone's mind.

File: pkg/kubelet/network/exec/exec.go (hunk 1 above)
Comment (thockin) [inline pkg/kubelet/network/exec/exec.go]: IF we were going to do this, I would argue that we would collect stdout and return that, and you could log to stderr instead.

Comment (nickethier) [inline pkg/kubelet/network/exec/exec.go]: :+1: agree will fix and push

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (thockin) [inline pkg/kubelet/dockertools/manager.go]: First, env vars in general are a really bad way of sharing information.  Second, we can not just assume control of an env var - the general pattern for these things is that we offer an API by which a user can insert our data into their own env var explicitly.

**Comment by thockin** [comment]
Nick,

This is sort of a big idea in a small package.  I'd like to talk about the idea more than the implementation.

You want to share some information with containers - what is that information?  What structure does it take?  How universal or specific is that information?  Does that information ever change? 

The idea of adding metadata on a link-local interface is something that has been dicsussed, but not something to enter lightly.  Once we expose it, we are stuck with it for a long time.

**Comment by nickethier** [comment]
Yes I can see now the implications are perhaps larger than I first anticipated. 

I'd like to expose this metadata at application runtime and it does not change. It is environment data about the networks the pod has been configured for, coordinates for what cluster it's running on, etc.

All of this is stored in annotations or labels so I suppose the end goal is to expose the pods objectmeta without having to make a network call to the apiserver. In my case none of this metadata changes after the network plugin has finished executing.

**Comment by thockin** [comment]
In your case none of it changes.  But it is plausible that it might change
in someone else's case?

Can you describe the structure of the data?  Maybe we can find a useful
struct that we can carry around and expose through some API (a volume, env,
HTTP)

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/kubelet/dockertools/manager.go]: +1 to what @thockin wrote. We don't force particular env var names onto users.

See: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-guide/downward-api.md

cc @pmorie

**Comment by nickethier** [comment]
I'll do my best, however I'm building abstractions on top of k8s. so apologies if I miss some context. Feel free to ask questions! (I love talking about this project to anyone who will listen.)

So I've been working on building a tool at my dayjob that provides an interface to manage multiple k8s clusters. It provides a few higher level abstractions that hide the notion of a specific cluster. There are some coordinates from this abstraction that I currently expose to pods as environment variables, and store in the PodSpec's annotations as:
 
```
jive.com/project -> "foo"
jive.com/app/name -> "bar"
jive.com/app/version -> "1"
```

This abstraction also allows you to specify which network(s) you'd like the pod to be on. Networks are just string keys that map back to an uid of a Network object. The network plugin pulls the list of networks out of the Pod's annotation and looks up the associated Network objects in the higher abstraction which gives it information on how to build the required interface(s) and route table. We make use of the following annotations for this:

```
jive.com/networks -> "<net1>,<net2>..."
jive.com/networks/<net1>/addresses -> "<ip1>" 
jive.com/networks/<net1>/vnid -> "1234"
```

For example I don't make use of k8s Services currently because they don't fit the multi-cluster model very well, so a Pod may need access to the "public" network to get an advertised public IP and access to the "internal" network to talk to other services. 

Other metadata may include geographical coordinates.

```
jive.com/cluster -> "nyc-1a"
jive.com/country -> "US"
jive.com/region -> us-east
jive.com/tags -> "nyc,prod,usa,etc."
jive.com/geo -> "00.00000,00.00000"
```

This "structure" if you can call it that, is evolving as well as I continue development on the tool. 

**Warning, my opinion**:
Once a Pod has been started, there shouldn't be a scenario where its metadata should change. Pods are to be immutable there for if something does need to be changed then a new Pod should be created, whether manually or by changing a replication controller.

**Comment by nickethier** [comment]
I'll also add that I completely agree that environment variables are not the way to do this and that discussing the idea is more important for this pull.

**Comment by dchen1107** [comment]
@nickethier based on the above discussion, I am going to close the pr. Please file an issue separately with detail requirements. Thanks!

Also please take a look at https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/proposals/federation.md Current proposal of Ubernetes doesn't really solve any of your issues, but what you required here might be one of requirements for that project.

**Pull request closed without merging**